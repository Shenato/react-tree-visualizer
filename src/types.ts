import React, { ReactElement } from 'react';

export type Options = {
  width?: number;

  boxHeight?: number;

  canvasPadding?: number;

  spaceBetweenColumns?: number;

  spaceBetweenRows?: number;

  connectorColor?: string;

  connectorColorHighlight?: string;

  connectorColorActive?: string;

  connectorLineThickness?: number;

  roundHeader?: {
    isShown?: boolean;
    height?: number;
    marginBottom?: number;
    fontSize?: number;
    fontColor?: string;
    backgroundColor?: string;
    fontFamily?: string;
  };

  roundSeparatorWidth?: number;

  lineInfo?: {
    separation?: number;
    homeVisitorSpread?: number;
  };

  horizontalOffset?: number;

  wonBywalkOverText?: string;

  lostByNoShowText?: string;
};

export type ComputedOptions = Options & {
  rowHeight?: number;

  columnWidth?: number;
};

export type CommonTreeProps = {
  svgWrapper?: (props: {
    bracketWidth: number;
    bracketHeight: number;
    startAt: number[];
    children: ReactElement;
  }) => React.ReactElement;
  options?: { style: Options };
};

export type RenderItem<T> = {
  id: number;
  uniqueId: string;
  parentId?: number;
  parentUniqueId: string;
  y: number;
  treePath: number[];
  data: T;
};

export type RenderMatrix<T> = RenderItem<T>[][];

export type Tree<T> = { id: string; data: T; children: Tree<T>[] };

export type ItemComponentProps<T> = {
  item: T;
  isHighlighted: boolean;
  isActive: boolean;
  isCollapsed: boolean;
  connectorColor: string;
  computedStyles: ComputedOptions;
  onMouseEnter: () => void;
  onMouseLeave: () => void;
  toggleCollapse: () => void;
};

export type GeneralTreeProps<T> = CommonTreeProps & {
  itemComponent: (props: ItemComponentProps<T>) => JSX.Element;
  activeItemIds?: string[];
  currentRound?: string;
  tree: Tree<T>;
};

export type DebugItemDataType = {
  type?: number;
  id: number;
  disabled?: boolean;
  hidden?: boolean;
  scrollAxis?: number;
};
