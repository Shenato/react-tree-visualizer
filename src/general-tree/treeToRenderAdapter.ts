import { RenderMatrix, Tree } from '../types';

/**
 * Return data should look like this
 *  [
 *    [ 1st Column ]
 *    [ 2nd Column ]
 *    [ 3rd Column ]
 *    [ 4th Column ]
 *  ]
 *
 */
export function generateRawRenderData<T>(
  generalTree: Tree<T>,
  collapsedIds: string[] = [],
  depth: number = 0,
  verticalStartingPoint = 0,
  parentUniqueId: string | undefined = undefined,
  parentId: number | undefined = undefined,
  id: number = 0,
  treePath: number[] = [],
  ancestryIds: string[] = [],
  result: RenderMatrix<T> = [[]]
): {
  verticalStartingPoint: number;
  id: number;
  newResult: RenderMatrix<T>;
} {
  let currentVerticalStartingPoint = verticalStartingPoint;
  let currentResult = result;
  let currentIterationId = id;

  const { children, data, id: currentItemUniqueId, ...rest } = generalTree;
  const item = {
    ...rest,
    uniqueId: currentItemUniqueId,
    id: currentIterationId,
    parentUniqueId,
    parentId,
    y: currentVerticalStartingPoint,
    ancestryIds,
    matrixPosition: [
      currentResult.length - 1,
      currentResult[depth]?.length ?? 0,
    ],
    treePath,
    data,
  };

  currentResult[depth] = currentResult[depth]?.length
    ? [...currentResult[depth], item]
    : [item];

  if (children.length === 0 || collapsedIds.includes(currentItemUniqueId)) {
    return {
      verticalStartingPoint: currentVerticalStartingPoint + 1,
      newResult: currentResult,
      id: currentIterationId,
    };
  }

  for (let i = 0; i < children.length; i++) {
    const {
      newResult,
      verticalStartingPoint: resultingVerticalStartingPoint,
      id: resultId,
    } = generateRawRenderData(
      children[i],
      collapsedIds,
      depth + 1,
      currentVerticalStartingPoint,
      currentItemUniqueId,
      id,
      currentIterationId + 1,
      [...treePath, i],
      [...ancestryIds, currentItemUniqueId],
      currentResult
    );
    currentIterationId = resultId;
    currentVerticalStartingPoint = resultingVerticalStartingPoint;
    currentResult = newResult;
  }

  return {
    verticalStartingPoint: currentVerticalStartingPoint,
    newResult: currentResult,
    id: currentIterationId,
  };
}

export function generateRenderData<T>(
  tree: Tree<T>[],
  collpasedItemIds?: string[]
) {
  const { newResult: renderDataRawUpdated } = generateRawRenderData<T>(
    {
      id: '',
      data: { id: 0, ref: false, type: 0, isFocused: false, shortcuts: [] },
      children: tree,
    },
    collpasedItemIds
  );
  const [, ...renderMatrix] = renderDataRawUpdated;

  return renderMatrix;
}
