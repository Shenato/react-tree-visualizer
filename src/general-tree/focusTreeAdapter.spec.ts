import { DebugItemDataType } from 'types';
import { generateRenderData } from './treeToRenderAdapter';
import { simpleTree, tinyTree } from '../mock-data/simple-tree';

describe('given a real example general tree', () => {
  it('converts it to render data', () => {
    const newResult = generateRenderData<DebugItemDataType>(simpleTree);

    expect(newResult).toMatchSnapshot();
  });
});

describe('given a bigger example', () => {
  it('converts it to render data', () => {
    const newResult = generateRenderData<DebugItemDataType>(tinyTree);
    expect(newResult).toMatchSnapshot();
  });
});

describe('given a bigger example', () => {
  it('converts it to render data with collapsed children', () => {
    const newResult = generateRenderData<DebugItemDataType>(tinyTree, ['4']);
    expect(newResult).toMatchSnapshot();
  });
});
