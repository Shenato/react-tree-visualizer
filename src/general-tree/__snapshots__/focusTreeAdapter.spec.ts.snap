// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given a bigger example converts it to render data 1`] = `
Array [
  Array [
    Object {
      "ancestryIds": Array [
        "0",
      ],
      "data": Object {
        "disabled": false,
        "hidden": false,
        "id": 1,
        "type": 1,
      },
      "id": 1,
      "matrixPosition": Array [
        0,
        0,
      ],
      "parentId": 0,
      "parentUniqueId": "0",
      "treePath": Array [
        0,
      ],
      "uniqueId": "1",
      "y": 0,
    },
    Object {
      "ancestryIds": Array [
        "0",
      ],
      "data": Object {
        "disabled": false,
        "hidden": false,
        "id": 4,
        "type": 2,
      },
      "id": 4,
      "matrixPosition": Array [
        2,
        1,
      ],
      "parentId": 0,
      "parentUniqueId": "0",
      "treePath": Array [
        1,
      ],
      "uniqueId": "4",
      "y": 2,
    },
  ],
  Array [
    Object {
      "ancestryIds": Array [
        "0",
        "1",
      ],
      "data": Object {
        "disabled": false,
        "hidden": false,
        "id": 2,
        "type": 1,
      },
      "id": 2,
      "matrixPosition": Array [
        1,
        0,
      ],
      "parentId": 1,
      "parentUniqueId": "1",
      "treePath": Array [
        0,
        0,
      ],
      "uniqueId": "2",
      "y": 0,
    },
    Object {
      "ancestryIds": Array [
        "0",
        "1",
      ],
      "data": Object {
        "disabled": false,
        "hidden": false,
        "id": 3,
        "type": 2,
      },
      "id": 3,
      "matrixPosition": Array [
        2,
        1,
      ],
      "parentId": 1,
      "parentUniqueId": "1",
      "treePath": Array [
        0,
        1,
      ],
      "uniqueId": "3",
      "y": 1,
    },
    Object {
      "ancestryIds": Array [
        "0",
        "4",
      ],
      "data": Object {
        "disabled": false,
        "hidden": false,
        "id": 5,
        "type": 1,
      },
      "id": 5,
      "matrixPosition": Array [
        2,
        2,
      ],
      "parentId": 4,
      "parentUniqueId": "4",
      "treePath": Array [
        1,
        0,
      ],
      "uniqueId": "5",
      "y": 2,
    },
  ],
]
`;

exports[`given a bigger example converts it to render data with collapsed children 1`] = `
Array [
  Array [
    Object {
      "ancestryIds": Array [
        "0",
      ],
      "data": Object {
        "disabled": false,
        "hidden": false,
        "id": 1,
        "type": 1,
      },
      "id": 1,
      "matrixPosition": Array [
        0,
        0,
      ],
      "parentId": 0,
      "parentUniqueId": "0",
      "treePath": Array [
        0,
      ],
      "uniqueId": "1",
      "y": 0,
    },
    Object {
      "ancestryIds": Array [
        "0",
      ],
      "data": Object {
        "disabled": false,
        "hidden": false,
        "id": 4,
        "type": 2,
      },
      "id": 4,
      "matrixPosition": Array [
        2,
        1,
      ],
      "parentId": 0,
      "parentUniqueId": "0",
      "treePath": Array [
        1,
      ],
      "uniqueId": "4",
      "y": 2,
    },
  ],
  Array [
    Object {
      "ancestryIds": Array [
        "0",
        "1",
      ],
      "data": Object {
        "disabled": false,
        "hidden": false,
        "id": 2,
        "type": 1,
      },
      "id": 2,
      "matrixPosition": Array [
        1,
        0,
      ],
      "parentId": 1,
      "parentUniqueId": "1",
      "treePath": Array [
        0,
        0,
      ],
      "uniqueId": "2",
      "y": 0,
    },
    Object {
      "ancestryIds": Array [
        "0",
        "1",
      ],
      "data": Object {
        "disabled": false,
        "hidden": false,
        "id": 3,
        "type": 2,
      },
      "id": 3,
      "matrixPosition": Array [
        2,
        1,
      ],
      "parentId": 1,
      "parentUniqueId": "1",
      "treePath": Array [
        0,
        1,
      ],
      "uniqueId": "3",
      "y": 1,
    },
  ],
]
`;

exports[`given a real example general tree converts it to render data 1`] = `
Array [
  Array [
    Object {
      "ancestryIds": Array [
        "0",
      ],
      "data": Object {
        "disabled": false,
        "hidden": false,
        "id": 4,
        "type": 1,
      },
      "id": 1,
      "matrixPosition": Array [
        0,
        0,
      ],
      "parentId": 0,
      "parentUniqueId": "0",
      "treePath": Array [
        0,
      ],
      "uniqueId": "4",
      "y": 0,
    },
  ],
  Array [
    Object {
      "ancestryIds": Array [
        "0",
        "4",
      ],
      "data": Object {
        "disabled": false,
        "hidden": false,
        "id": 14,
        "type": 1,
      },
      "id": 2,
      "matrixPosition": Array [
        1,
        0,
      ],
      "parentId": 1,
      "parentUniqueId": "4",
      "treePath": Array [
        0,
        0,
      ],
      "uniqueId": "14",
      "y": 0,
    },
    Object {
      "ancestryIds": Array [
        "0",
        "4",
      ],
      "data": Object {
        "disabled": false,
        "hidden": false,
        "id": 15,
        "type": 1,
      },
      "id": 4,
      "matrixPosition": Array [
        3,
        1,
      ],
      "parentId": 1,
      "parentUniqueId": "4",
      "treePath": Array [
        0,
        1,
      ],
      "uniqueId": "15",
      "y": 1,
    },
  ],
  Array [
    Object {
      "ancestryIds": Array [
        "0",
        "4",
        "14",
      ],
      "data": Object {
        "id": 45,
        "type": 0,
      },
      "id": 3,
      "matrixPosition": Array [
        2,
        0,
      ],
      "parentId": 2,
      "parentUniqueId": "14",
      "treePath": Array [
        0,
        0,
        0,
      ],
      "uniqueId": "45",
      "y": 0,
    },
    Object {
      "ancestryIds": Array [
        "0",
        "4",
        "15",
      ],
      "data": Object {
        "disabled": false,
        "hidden": false,
        "id": 17,
        "scrollAxis": 1,
        "type": 1,
      },
      "id": 5,
      "matrixPosition": Array [
        3,
        1,
      ],
      "parentId": 4,
      "parentUniqueId": "15",
      "treePath": Array [
        0,
        1,
        0,
      ],
      "uniqueId": "17",
      "y": 1,
    },
    Object {
      "ancestryIds": Array [
        "0",
        "4",
        "15",
      ],
      "data": Object {
        "disabled": false,
        "hidden": false,
        "id": 19,
        "scrollAxis": 1,
        "type": 1,
      },
      "id": 14,
      "matrixPosition": Array [
        4,
        2,
      ],
      "parentId": 4,
      "parentUniqueId": "15",
      "treePath": Array [
        0,
        1,
        1,
      ],
      "uniqueId": "19",
      "y": 9,
    },
  ],
  Array [
    Object {
      "ancestryIds": Array [
        "0",
        "4",
        "15",
        "17",
      ],
      "data": Object {
        "id": 22,
        "type": 0,
      },
      "id": 6,
      "matrixPosition": Array [
        3,
        0,
      ],
      "parentId": 5,
      "parentUniqueId": "17",
      "treePath": Array [
        0,
        1,
        0,
        0,
      ],
      "uniqueId": "22",
      "y": 1,
    },
    Object {
      "ancestryIds": Array [
        "0",
        "4",
        "15",
        "17",
      ],
      "data": Object {
        "id": 23,
        "type": 0,
      },
      "id": 7,
      "matrixPosition": Array [
        4,
        1,
      ],
      "parentId": 5,
      "parentUniqueId": "17",
      "treePath": Array [
        0,
        1,
        0,
        1,
      ],
      "uniqueId": "23",
      "y": 2,
    },
    Object {
      "ancestryIds": Array [
        "0",
        "4",
        "15",
        "17",
      ],
      "data": Object {
        "id": 24,
        "type": 0,
      },
      "id": 8,
      "matrixPosition": Array [
        4,
        2,
      ],
      "parentId": 5,
      "parentUniqueId": "17",
      "treePath": Array [
        0,
        1,
        0,
        2,
      ],
      "uniqueId": "24",
      "y": 3,
    },
    Object {
      "ancestryIds": Array [
        "0",
        "4",
        "15",
        "17",
      ],
      "data": Object {
        "id": 25,
        "type": 0,
      },
      "id": 9,
      "matrixPosition": Array [
        4,
        3,
      ],
      "parentId": 5,
      "parentUniqueId": "17",
      "treePath": Array [
        0,
        1,
        0,
        3,
      ],
      "uniqueId": "25",
      "y": 4,
    },
    Object {
      "ancestryIds": Array [
        "0",
        "4",
        "15",
        "17",
      ],
      "data": Object {
        "id": 26,
        "type": 0,
      },
      "id": 10,
      "matrixPosition": Array [
        4,
        4,
      ],
      "parentId": 5,
      "parentUniqueId": "17",
      "treePath": Array [
        0,
        1,
        0,
        4,
      ],
      "uniqueId": "26",
      "y": 5,
    },
    Object {
      "ancestryIds": Array [
        "0",
        "4",
        "15",
        "17",
      ],
      "data": Object {
        "id": 27,
        "type": 0,
      },
      "id": 11,
      "matrixPosition": Array [
        4,
        5,
      ],
      "parentId": 5,
      "parentUniqueId": "17",
      "treePath": Array [
        0,
        1,
        0,
        5,
      ],
      "uniqueId": "27",
      "y": 6,
    },
    Object {
      "ancestryIds": Array [
        "0",
        "4",
        "15",
        "17",
      ],
      "data": Object {
        "id": 28,
        "type": 0,
      },
      "id": 12,
      "matrixPosition": Array [
        4,
        6,
      ],
      "parentId": 5,
      "parentUniqueId": "17",
      "treePath": Array [
        0,
        1,
        0,
        6,
      ],
      "uniqueId": "28",
      "y": 7,
    },
    Object {
      "ancestryIds": Array [
        "0",
        "4",
        "15",
        "17",
      ],
      "data": Object {
        "id": 29,
        "type": 0,
      },
      "id": 13,
      "matrixPosition": Array [
        4,
        7,
      ],
      "parentId": 5,
      "parentUniqueId": "17",
      "treePath": Array [
        0,
        1,
        0,
        7,
      ],
      "uniqueId": "29",
      "y": 8,
    },
    Object {
      "ancestryIds": Array [
        "0",
        "4",
        "15",
        "19",
      ],
      "data": Object {
        "id": 21,
        "type": 0,
      },
      "id": 15,
      "matrixPosition": Array [
        4,
        8,
      ],
      "parentId": 14,
      "parentUniqueId": "19",
      "treePath": Array [
        0,
        1,
        1,
        0,
      ],
      "uniqueId": "21",
      "y": 9,
    },
    Object {
      "ancestryIds": Array [
        "0",
        "4",
        "15",
        "19",
      ],
      "data": Object {
        "disabled": true,
        "id": 20,
        "type": 0,
      },
      "id": 16,
      "matrixPosition": Array [
        4,
        9,
      ],
      "parentId": 14,
      "parentUniqueId": "19",
      "treePath": Array [
        0,
        1,
        1,
        1,
      ],
      "uniqueId": "20",
      "y": 10,
    },
    Object {
      "ancestryIds": Array [
        "0",
        "4",
        "15",
        "19",
      ],
      "data": Object {
        "id": 30,
        "type": 0,
      },
      "id": 17,
      "matrixPosition": Array [
        4,
        10,
      ],
      "parentId": 14,
      "parentUniqueId": "19",
      "treePath": Array [
        0,
        1,
        1,
        2,
      ],
      "uniqueId": "30",
      "y": 11,
    },
    Object {
      "ancestryIds": Array [
        "0",
        "4",
        "15",
        "19",
      ],
      "data": Object {
        "id": 31,
        "type": 0,
      },
      "id": 18,
      "matrixPosition": Array [
        4,
        11,
      ],
      "parentId": 14,
      "parentUniqueId": "19",
      "treePath": Array [
        0,
        1,
        1,
        3,
      ],
      "uniqueId": "31",
      "y": 12,
    },
    Object {
      "ancestryIds": Array [
        "0",
        "4",
        "15",
        "19",
      ],
      "data": Object {
        "disabled": true,
        "id": 32,
        "type": 0,
      },
      "id": 19,
      "matrixPosition": Array [
        4,
        12,
      ],
      "parentId": 14,
      "parentUniqueId": "19",
      "treePath": Array [
        0,
        1,
        1,
        4,
      ],
      "uniqueId": "32",
      "y": 13,
    },
    Object {
      "ancestryIds": Array [
        "0",
        "4",
        "15",
        "19",
      ],
      "data": Object {
        "disabled": true,
        "id": 33,
        "type": 0,
      },
      "id": 20,
      "matrixPosition": Array [
        4,
        13,
      ],
      "parentId": 14,
      "parentUniqueId": "19",
      "treePath": Array [
        0,
        1,
        1,
        5,
      ],
      "uniqueId": "33",
      "y": 14,
    },
    Object {
      "ancestryIds": Array [
        "0",
        "4",
        "15",
        "19",
      ],
      "data": Object {
        "disabled": true,
        "id": 34,
        "type": 0,
      },
      "id": 21,
      "matrixPosition": Array [
        4,
        14,
      ],
      "parentId": 14,
      "parentUniqueId": "19",
      "treePath": Array [
        0,
        1,
        1,
        6,
      ],
      "uniqueId": "34",
      "y": 15,
    },
    Object {
      "ancestryIds": Array [
        "0",
        "4",
        "15",
        "19",
      ],
      "data": Object {
        "id": 35,
        "type": 0,
      },
      "id": 22,
      "matrixPosition": Array [
        4,
        15,
      ],
      "parentId": 14,
      "parentUniqueId": "19",
      "treePath": Array [
        0,
        1,
        1,
        7,
      ],
      "uniqueId": "35",
      "y": 16,
    },
    Object {
      "ancestryIds": Array [
        "0",
        "4",
        "15",
        "19",
      ],
      "data": Object {
        "id": 36,
        "type": 0,
      },
      "id": 23,
      "matrixPosition": Array [
        4,
        16,
      ],
      "parentId": 14,
      "parentUniqueId": "19",
      "treePath": Array [
        0,
        1,
        1,
        8,
      ],
      "uniqueId": "36",
      "y": 17,
    },
    Object {
      "ancestryIds": Array [
        "0",
        "4",
        "15",
        "19",
      ],
      "data": Object {
        "id": 37,
        "type": 0,
      },
      "id": 24,
      "matrixPosition": Array [
        4,
        17,
      ],
      "parentId": 14,
      "parentUniqueId": "19",
      "treePath": Array [
        0,
        1,
        1,
        9,
      ],
      "uniqueId": "37",
      "y": 18,
    },
    Object {
      "ancestryIds": Array [
        "0",
        "4",
        "15",
        "19",
      ],
      "data": Object {
        "id": 38,
        "type": 0,
      },
      "id": 25,
      "matrixPosition": Array [
        4,
        18,
      ],
      "parentId": 14,
      "parentUniqueId": "19",
      "treePath": Array [
        0,
        1,
        1,
        10,
      ],
      "uniqueId": "38",
      "y": 19,
    },
    Object {
      "ancestryIds": Array [
        "0",
        "4",
        "15",
        "19",
      ],
      "data": Object {
        "id": 39,
        "type": 0,
      },
      "id": 26,
      "matrixPosition": Array [
        4,
        19,
      ],
      "parentId": 14,
      "parentUniqueId": "19",
      "treePath": Array [
        0,
        1,
        1,
        11,
      ],
      "uniqueId": "39",
      "y": 20,
    },
    Object {
      "ancestryIds": Array [
        "0",
        "4",
        "15",
        "19",
      ],
      "data": Object {
        "id": 40,
        "type": 0,
      },
      "id": 27,
      "matrixPosition": Array [
        4,
        20,
      ],
      "parentId": 14,
      "parentUniqueId": "19",
      "treePath": Array [
        0,
        1,
        1,
        12,
      ],
      "uniqueId": "40",
      "y": 21,
    },
    Object {
      "ancestryIds": Array [
        "0",
        "4",
        "15",
        "19",
      ],
      "data": Object {
        "id": 41,
        "type": 0,
      },
      "id": 28,
      "matrixPosition": Array [
        4,
        21,
      ],
      "parentId": 14,
      "parentUniqueId": "19",
      "treePath": Array [
        0,
        1,
        1,
        13,
      ],
      "uniqueId": "41",
      "y": 22,
    },
    Object {
      "ancestryIds": Array [
        "0",
        "4",
        "15",
        "19",
      ],
      "data": Object {
        "id": 42,
        "type": 0,
      },
      "id": 29,
      "matrixPosition": Array [
        4,
        22,
      ],
      "parentId": 14,
      "parentUniqueId": "19",
      "treePath": Array [
        0,
        1,
        1,
        14,
      ],
      "uniqueId": "42",
      "y": 23,
    },
    Object {
      "ancestryIds": Array [
        "0",
        "4",
        "15",
        "19",
      ],
      "data": Object {
        "id": 43,
        "type": 0,
      },
      "id": 30,
      "matrixPosition": Array [
        4,
        23,
      ],
      "parentId": 14,
      "parentUniqueId": "19",
      "treePath": Array [
        0,
        1,
        1,
        15,
      ],
      "uniqueId": "43",
      "y": 24,
    },
    Object {
      "ancestryIds": Array [
        "0",
        "4",
        "15",
        "19",
      ],
      "data": Object {
        "id": 44,
        "type": 0,
      },
      "id": 31,
      "matrixPosition": Array [
        4,
        24,
      ],
      "parentId": 14,
      "parentUniqueId": "19",
      "treePath": Array [
        0,
        1,
        1,
        16,
      ],
      "uniqueId": "44",
      "y": 25,
    },
  ],
]
`;
